From 59b7259ab53a4fbacaff9050ecec559b5f534d2e Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@freescale.com>
Date: Fri, 14 Oct 2016 11:44:34 +0800
Subject: [PATCH 1208/5242] MLK-13224: ASoC: imx-hdmi-dma: fix glitch noise
 issue in long time playback

commit  e7b9937d73def00a18317149815bccd469721154 from
https://source.codeaurora.org/external/imx/linux-imx.git

The calculation "runtime->status->hw_ptr * (runtime->frame_bits / 8)" may
exceed the integer scope, then appl_bytes is no correct.
This patch is to fix this issue.

Signed-off-by: Shengjiu Wang <shengjiu.wang@freescale.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/imx-hdmi-dma.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/sound/soc/fsl/imx-hdmi-dma.c b/sound/soc/fsl/imx-hdmi-dma.c
index e3b81e5..3410d03 100644
--- a/sound/soc/fsl/imx-hdmi-dma.c
+++ b/sound/soc/fsl/imx-hdmi-dma.c
@@ -1,7 +1,7 @@
 /*
  * imx-hdmi-dma.c  --  HDMI DMA driver for ALSA Soc Audio Layer
  *
- * Copyright (C) 2011-2014 Freescale Semiconductor, Inc.
+ * Copyright (C) 2011-2016 Freescale Semiconductor, Inc.
  *
  * based on imx-pcm-dma-mx2.c
  * Copyright 2009 Sascha Hauer <s.hauer@pengutronix.de>
@@ -458,7 +458,7 @@ static void hdmi_dma_data_copy(struct snd_pcm_substream *substream,
 	if (runtime->access != SNDRV_PCM_ACCESS_MMAP_INTERLEAVED)
 		return;
 
-	appl_bytes =  runtime->status->hw_ptr * (runtime->frame_bits / 8);
+	appl_bytes =  (runtime->status->hw_ptr % (priv->buffer_bytes/(runtime->frame_bits/8))) * (runtime->frame_bits/8);
 	if (type == 'p')
 		appl_bytes += 2 * priv->period_bytes;
 	offset = appl_bytes % priv->buffer_bytes;
-- 
1.7.9.5

